internal/validate/string.go

```diff
@@ -106,3 +107,35 @@ func inRange(ranges []CharRange, c rune) bool {
 	}
 	return false
 }
+
+// Enum returns a validation rule that checks that value is one of the allowed
+// strings.
+func Enum(name string, value string, allowed []string) ValidationRule {
+	return enum{Name: name, Value: value, Allowed: allowed}
+}
+
+type enum struct {
+	Name    string
+	Value   string
+	Allowed []string
+}
+
+func (e enum) Validate() *Failure {
+	if e.Value == "" {
```

Comment:
Maybe this should be called `OptionalEnum`, it isnt super obvious that this will return valid for an empty value. That is the behavior we want in the providers case, but maybe not in general.

Reply:
Good call out! This is different from our previous library. 

I've been making all of the validation rules optional, for example:
* a `StringRule` with a minLength > 0 is still valid if the string is the empty string
* an int of `0` is still valid, even with an `IntRule` with a min > 0
* an email address is still valid if it's the empty string

Since we have a `Required` validation rule, it seems like we already have a good way of making a field required, apply both the required rule and some other rule. 

So it seemed to me like all rules should default to optional, which means we probably don't need to name them that way.  

But I think it might be good to call that out in the docs. Maybe the godoc for `ValidationRule` should say that?

Reply:
That seems reasonable to me 👍